File: sources/06-hash-table/01-hash-table-direct-with-reservation/java/Main.java
class Hash {
    int table[];
    int m1, m2, m, reservation;

    public Hash() {
        this.m1 = 13;
        this.m2 = 7;
    }

    public Hash(int m1, int m2) {
        this.m1 = m1;
        this.m2 = m2;
        this.m = m1 + m2;
        this.table = new int[this.m];

        for (int i = 0; i < m1; i++) {
           table[i] = -1;
        }

        this.reservation = 0;
    }

    public int hash(int value) {
        return value % m1;
    }

    public void insert(int value) {
        int i = hash(value);

        if (value == -1) {
            System.out.println("Favor inserir um valor.");
        } else if (table[i] == -1) {
            table[i] = value;
        } else if (reservation < m2) {
            table[m1 + reservation] = value;
            reservation++;
        } else {
            System.out.println("Erro ao inserir elemento na tabela hash.");
        }
    }

    public boolean search(int value) {
        int i = hash(value);
        boolean found = false;

        if (table[i] == value) {
            found = true;
        } else if (table[i] != -1) {
            for (int j = 0; j < reservation; j++) {
                if (table[m1 + j] == value) {
                    found = true;
                    j = reservation;
                }
            }
        }

        return found;
    }

    public boolean removal(int value) {
        int i = hash(value);
        boolean removed = false;

        if (table[i] == value) {
            table[i] = -2;
            removed = true;
        } else if (table[i] != -1) {
            for (int j = 0; j < reservation; j++) {
                if (table[m1 + j] == value) {
                    table[m1 + j] = -2;
                    removed = true;
                    j = reservation;
                }
            }
        }

        return removed;
    }

    public void show() {
        System.out.println("Índice\tValor");

        for (int i = 0; i < m; i++) {
            System.out.println(i + "\t" + table[i]);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Hash hashTable = new Hash(13, 7);

        System.out.println("Inserindo valores...");
        hashTable.insert(10);
        hashTable.insert(23);
        hashTable.insert(36);
        hashTable.insert(49);
        hashTable.insert(15);
        hashTable.insert(64);

        System.out.println("\nTabela Hash após inserções:");
        hashTable.show();

        System.out.println("\nPesquisando valores...");
        System.out.println("Valor 23 encontrado? " + hashTable.search(23));
        System.out.println("Valor 99 encontrado? " + hashTable.search(99));

        System.out.println("\nRemovendo valores...");
        System.out.println("Valor 23 removido? " + hashTable.removal(23));
        System.out.println("Valor 99 removido? " + hashTable.removal(99));

        System.out.println("\nTabela Hash após remoções:");
        hashTable.show();
    }
}

File: sources/06-hash-table/01-hash-table-direct-with-reservation/c/Main.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct {
    int *table;
    int m1, m2, m, reservation;
} Hash;

void initHash(Hash *hash, int m1, int m2) {
    hash->m1 = m1;
    hash->m2 = m2;
    hash->m = m1 + m2;
    hash->reservation = 0;

    hash->table = (int *) malloc(hash->m * sizeof(int));
    
    for (int i = 0; i < hash->m; i++) {
        hash->table[i] = -1;
    }
}

int h(Hash *hash, int value) {
    return value % hash->m1;
}

void insert(Hash *hash, int value) {
    int i = h(hash, value);

    if (value == -1) {
        printf("Favor inserir um valor.\n");
    } else if (hash->table[i] == -1) {
        hash->table[i] = value;
    } else if (hash->reservation < hash->m2) {
        hash->table[hash->m1 + hash->reservation] = value;
        hash->reservation++;
    } else {
        printf("Erro ao inserir elemento na tabela hash.\n");
    }
}

bool search(Hash *hash, int value) {
    int i = h(hash, value);

    if (hash->table[i] == value) {
        return true;
    }

    for (int j = 0; j < hash->reservation; j++) {
        if (hash->table[hash->m1 + j] == value) {
            return true;
        }
    }

    return false;
}

bool removal(Hash *hash, int value) {
    int i = h(hash, value);

    if (hash->table[i] == value) {
        hash->table[i] = -2;
        return true;
    }

    for (int j = 0; j < hash->reservation; j++) {
        if (hash->table[hash->m1 + j] == value) {
            hash->table[hash->m1 + j] = -2;
            return true;
        }
    }

    return false;
}

void show(Hash *hash) {
    printf("Índice\tValor\n");
    for (int i = 0; i < hash->m; i++) {
        printf("%d\t%d\n", i, hash->table[i]);
    }
}

void freeHash(Hash *hash) {
    free(hash->table);
}

int main() {
    Hash hashTable;
    initHash(&hashTable, 13, 7);

    printf("Inserindo valores...\n");
    insert(&hashTable, 10);
    insert(&hashTable, 23);
    insert(&hashTable, 36);
    insert(&hashTable, 49);
    insert(&hashTable, 15);
    insert(&hashTable, 64);

    printf("\nTabela Hash após inserções:\n");
    show(&hashTable);

    printf("\nPesquisando valores...\n");
    printf("Valor 23 encontrado? %s\n", search(&hashTable, 23) ? "true" : "false");
    printf("Valor 99 encontrado? %s\n", search(&hashTable, 99) ? "true" : "false");

    printf("\nRemovendo valores...\n");
    printf("Valor 23 removido? %s\n", removal(&hashTable, 23) ? "true" : "false");
    printf("Valor 99 removido? %s\n", removal(&hashTable, 99) ? "true" : "false");

    printf("\nTabela Hash após remoções:\n");
    show(&hashTable);

    // Libera memória alocada para a tabela hash
    freeHash(&hashTable);

    return 0;
}